plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE TupleSections      #-}
+{-# LANGUAGE ViewPatterns       #-}
+
+module PlutusTx.DataMap where
+
+import PlutusTx.Builtins qualified as P
+import PlutusTx.Builtins.Internal qualified as BI
+import PlutusTx.IsData qualified as P
+import PlutusTx.Prelude hiding (null, toList)
+import PlutusTx.These
+
+import Prelude qualified as Haskell
+
+{- | An associative map implementation backed by `P.BuiltinData`.
+
+This map implementation has the following characteristics:
+
+  * The `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations are no-op.
+  * Other operations are slower than @PlutusTx.AssocMap.Map@, although equality
+    checks on keys can be faster due to `P.equalsData`.
+  * Many operations involve converting the keys and/or values to/from `P.BuiltinData`.
+
+Therefore this map implementation is likely a better choice than @PlutusTx.AssocMap.Map@
+if it is part of a data type defined using @asData@, and the key and value types
+have efficient `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations (e.g., they
+are primitive types or types defined using @asData@).
+-}
+newtype Map k a = Map P.BuiltinData
```

Comment:
A comment on whether there are any invariants such as orderness? (Stipulating the lack of orderness would also be helpful)

Reply:
I think the invariant is just no-duplicates, and not orderedness. I wonder if having it be ordered would help... then it wouldn't be a drop-in replacement for `AssocMap` since you would need `Ord`, I guess.

Reply:
Isn't the notion of invariant a bit tenuous here?  It would be easy to construct a data object that looks like a map but doesn't satisfy the invariants and stick it in a datum or something. You might even do that by accident if you're using some external tool to produce data.

Reply:
> Isn't the notion of invariant a bit tenuous here?

Either way, I just wanna know if there are any or not.

>  It would be easy to construct a data object that looks like a map but doesn't satisfy the invariants and stick it in a datum or something. You might even do that by accident if you're using some external tool to produce data.

A `Map` isn't necessarily taken verbatim from a datum. It may be produced by the validator itself or come from a datum but be properly sanitized, so I don't think having invariants is out of question. But maybe I misunderstand the intended use cases.

Reply:
> I think the invariant is just no-duplicates

It doesn't guarantee no-duplicates. This is really an association list, rather than a map, and should probably be named so.